feat(2135): Add resource limits/requests flags to helm init

ata18 opened the pull request:
This PR adds the following 4 additional flags to be used with helm init
command so that appropriate CPU/Memory limits can be set for tiller
container.
--tiller-cpu-limits
--tiller-cpu-requests
--tiller-memory-limits
--tiller-memory-requests

To maintain backwards compatibility, when none of these flags are set,
no limitrange will be added. When tiller-cpu-limits is specified but
not the tiller-cpu-requests, then tiller-cpu-requests is set to same
value as the tiller-cpu-limits. But not vice-versa. Same applies to
tiller-memory-limits and tiller-memory-requests.

Fixes: https://github.com/kubernetes/helm/issues/2135

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/helm/2632)
<!-- Reviewable:end -->

seh commented:
Review status: 0 of 5 files reviewed at latest revision, 40 unresolved discussions, some commit checks failed.

---

*[cmd/helm/init.go, line 83 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KntH60fApTsW-o0r0OW:-KntH60geZqnsdLknuT1:b-kbny38) ([raw file](https://github.com/kubernetes/helm/blob/10fd6f5c989e1284ef0e9276d989d5d6bc23828d/cmd/helm/init.go#L83)):*
> ```Go
> 	kubeClient     kubernetes.Interface
> 	serviceAccount string
> 	cpuLimits      string
> ```

s/cpuLimits/cpuLimit/

---

*[cmd/helm/init.go, line 84 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KntH8c2esDppUeUWtFU:-KntH8c2esDppUeUWtFV:bdc837o) ([raw file](https://github.com/kubernetes/helm/blob/10fd6f5c989e1284ef0e9276d989d5d6bc23828d/cmd/helm/init.go#L84)):*
> ```Go
> 	serviceAccount string
> 	cpuLimits      string
> 	cpuRequests    string
> ```

s/cpuRequests/cpuRequest/

---

*[cmd/helm/init.go, line 85 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KntHCtsuQwlhcQyc1zh:-KntHCtsuQwlhcQyc1zi:b-9ai5uw) ([raw file](https://github.com/kubernetes/helm/blob/10fd6f5c989e1284ef0e9276d989d5d6bc23828d/cmd/helm/init.go#L85)):*
> ```Go
> 	cpuLimits      string
> 	cpuRequests    string
> 	memLimits      string
> ```

s/memLimits/memoryLimit/

---

*[cmd/helm/init.go, line 86 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KntHEuLtaqAIGJodx6o:-KntHEuLtaqAIGJodx6p:bjt4yu0) ([raw file](https://github.com/kubernetes/helm/blob/10fd6f5c989e1284ef0e9276d989d5d6bc23828d/cmd/helm/init.go#L86)):*
> ```Go
> 	cpuRequests    string
> 	memLimits      string
> 	memRequests    string
> ```

s/memRequests/memoryRequest/

---

*[cmd/helm/init.go, line 128 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KntGYbOfFqmiTVxACDc:-KntGYbPt4EfvpAitY6z:b-phaws4) ([raw file](https://github.com/kubernetes/helm/blob/10fd6f5c989e1284ef0e9276d989d5d6bc23828d/cmd/helm/init.go#L128)):*
> tiller cpu limits

That should read "Tiller CPU limit."

---

*[cmd/helm/init.go, line 128 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KntGkrORzqIA7LNWbdd:-KntGkrORzqIA7LNWbde:bnd18lb) ([raw file](https://github.com/kubernetes/helm/blob/10fd6f5c989e1284ef0e9276d989d5d6bc23828d/cmd/helm/init.go#L128)):*
> ```Go
> 	f.StringVar(&i.serviceAccount, "service-account", "", "name of service account")
> 
> 	f.StringVar(&i.cpuLimits, "tiller-cpu-limits", "", "override tiller cpu limits")
> ```

The flag name should be singular: "tiller-cpu-limit."

---

*[cmd/helm/init.go, line 129 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KntGcBnyX5qNz5O-ilx:-KntGcBnyX5qNz5O-ily:b-phaws4) ([raw file](https://github.com/kubernetes/helm/blob/10fd6f5c989e1284ef0e9276d989d5d6bc23828d/cmd/helm/init.go#L129)):*
> tiller cpu limits

That should read "Tiller CPU limit."

---

*[cmd/helm/init.go, line 129 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KntGqjJ1F7tP03J4-Ys:-KntGqjJ1F7tP03J4-Yt:b-c67drh) ([raw file](https://github.com/kubernetes/helm/blob/10fd6f5c989e1284ef0e9276d989d5d6bc23828d/cmd/helm/init.go#L129)):*
> ```Go
> 
> 	f.StringVar(&i.cpuLimits, "tiller-cpu-limits", "", "override tiller cpu limits")
> 	f.StringVar(&i.cpuRequests, "tiller-cpu-requests", "", "override tiller cpu requests")
> ```

The flag name should be singular: "tiller-cpu-request."

---

*[cmd/helm/init.go, line 130 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KntGsYdFmBS3gInOMo5:-KntGsYdFmBS3gInOMo6:b-uysvoy) ([raw file](https://github.com/kubernetes/helm/blob/10fd6f5c989e1284ef0e9276d989d5d6bc23828d/cmd/helm/init.go#L130)):*
> ```Go
> 	f.StringVar(&i.cpuLimits, "tiller-cpu-limits", "", "override tiller cpu limits")
> 	f.StringVar(&i.cpuRequests, "tiller-cpu-requests", "", "override tiller cpu requests")
> 	f.StringVar(&i.memLimits, "tiller-memory-limits", "", "override tiller cpu limits")
> ```

The flag name should be singular: "tiller-memory-limit."

---

*[cmd/helm/init.go, line 130 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KntGxoOp-vsabA9s97s:-KntGxoOp-vsabA9s97t:b-f0aih1) ([raw file](https://github.com/kubernetes/helm/blob/10fd6f5c989e1284ef0e9276d989d5d6bc23828d/cmd/helm/init.go#L130)):*
> tiller cpu limits

That should read "Tiller memory limit."

---

*[cmd/helm/init.go, line 131 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KntGueeEfr6AM33Da4F:-KntGueeEfr6AM33Da4G:b-c67drh) ([raw file](https://github.com/kubernetes/helm/blob/10fd6f5c989e1284ef0e9276d989d5d6bc23828d/cmd/helm/init.go#L131)):*
> ```Go
> 	f.StringVar(&i.cpuRequests, "tiller-cpu-requests", "", "override tiller cpu requests")
> 	f.StringVar(&i.memLimits, "tiller-memory-limits", "", "override tiller cpu limits")
> 	f.StringVar(&i.memRequests, "tiller-memory-requests", "", "override tiller cpu limits")
> ```

The flag name should be singular: "tiller-cpu-request."

---

*[cmd/helm/init.go, line 131 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KntH-_PUbfACgzfVWE5:-KntH-_PUbfACgzfVWE6:b-phaws4) ([raw file](https://github.com/kubernetes/helm/blob/10fd6f5c989e1284ef0e9276d989d5d6bc23828d/cmd/helm/init.go#L131)):*
> tiller cpu limits

That should read "Tiller CPU limit."

---

*[cmd/helm/init.go, line 163 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KntHf2VUoh4-af4tTO7:-KntHf2WtRM0qGPDTrER:b-rrgpou) ([raw file](https://github.com/kubernetes/helm/blob/10fd6f5c989e1284ef0e9276d989d5d6bc23828d/cmd/helm/init.go#L163)):*
> ```Go
> func (i *initCmd) generateLimits() error {
> 	var err error
> 	if i.cpuLimits != "" {
> ```

Consider writing this as a walk over a table of the following:
```go
struct {
  spec string
  quantity *resource.Quantity
}
```
Sketching:
```go
for _, opt := range []struct{
  spec string
  quantity *resource.Quantity
}{
  {i.cpuLimit, &i.opts.CPULimit},
  {i.cpuRequest, &i.opts.CPURequest},
  {i.memoryRequest, &i.opts.MemoryRequest},
  {i.memoryLimit, &i.opts.MemoryLimit},
}{
  *opt.quantity, err = resource.ParseQuantity(opt.spec)
  if err != nil {
    return err
  }
}
```

---

*[cmd/helm/init_test.go, line 307 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KntIjRdUxrdSfuBFPXk:-KntIjRdUxrdSfuBFPXl:bco1vef) ([raw file](https://github.com/kubernetes/helm/blob/10fd6f5c989e1284ef0e9276d989d5d6bc23828d/cmd/helm/init_test.go#L307)):*
> ```Go
> 
> func TestInitCmd_resourceLimits(t *testing.T) {
> ```

Remove blank line.

---

*[cmd/helm/installer/install.go, line 120 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KntKI0wbylgFDs0puJw:-KntKI0wbylgFDs0puJx:bfzussl) ([raw file](https://github.com/kubernetes/helm/blob/10fd6f5c989e1284ef0e9276d989d5d6bc23828d/cmd/helm/installer/install.go#L120)):*
> ```Go
> }
> 
> func generateLimits(opts *Options) (v1.ResourceRequirements, bool) {
> ```

This function should be called something more like "generateResourceRequirements" or "generateRequirements."

---

*[cmd/helm/installer/install.go, line 121 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KntIy7M06RjlHmtCn0j:-KntIy7M06RjlHmtCn0k:bstzbol) ([raw file](https://github.com/kubernetes/helm/blob/10fd6f5c989e1284ef0e9276d989d5d6bc23828d/cmd/helm/installer/install.go#L121)):*
> ```Go
> 
> func generateLimits(opts *Options) (v1.ResourceRequirements, bool) {
> 	limits := map[v1.ResourceName]resource.Quantity{}
> ```

Use `make`, indicating that you'll have between up to two entries:
```go
make(map[v1.ResourceName]resource.Quantity, 2)
```

---

*[cmd/helm/installer/install.go, line 122 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KntJAPoZwc_aF378lB0:-KntJAPoZwc_aF378lB1:bstzbol) ([raw file](https://github.com/kubernetes/helm/blob/10fd6f5c989e1284ef0e9276d989d5d6bc23828d/cmd/helm/installer/install.go#L122)):*
> ```Go
> func generateLimits(opts *Options) (v1.ResourceRequirements, bool) {
> 	limits := map[v1.ResourceName]resource.Quantity{}
> 	requests := map[v1.ResourceName]resource.Quantity{}
> ```

Use `make`, indicating that you'll have between up to two entries:
```go
make(map[v1.ResourceName]resource.Quantity, 2)
```

---

*[cmd/helm/installer/install.go, line 125 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KntJcod-f2JKBLpl_Uq:-KntJcod-f2JKBLpl_Ur:b-r266rc) ([raw file](https://github.com/kubernetes/helm/blob/10fd6f5c989e1284ef0e9276d989d5d6bc23828d/cmd/helm/installer/install.go#L125)):*
> ```Go
> 	resources := v1.ResourceRequirements{}
> 	var resourcesEnabled bool
> 	if (opts.CPURequests != (resource.Quantity{})) || (opts.CPULimits != (resource.Quantity{})) ||
> ```

Eliminate this conditional, and set "resourcesEnabled" to true when you enter any of the four conditional statements that follow.

---

*[cmd/helm/installer/install.go, line 134 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KntJnBMBfnU5_-_KzS9:-KntJnBMBfnU5_-_KzSA:b2xs2x2) ([raw file](https://github.com/kubernetes/helm/blob/10fd6f5c989e1284ef0e9276d989d5d6bc23828d/cmd/helm/installer/install.go#L134)):*
> ```Go
> 			}
> 		}
> 		if opts.CPURequests != (resource.Quantity{}) {
> ```

Handle a request before a limit.
Move this up.

---

*[cmd/helm/installer/install.go, line 143 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KntJpVIBRZuWHWV_AtD:-KntJpVJGW7EGlcxD3l1:b2xs2x2) ([raw file](https://github.com/kubernetes/helm/blob/10fd6f5c989e1284ef0e9276d989d5d6bc23828d/cmd/helm/installer/install.go#L143)):*
> ```Go
> 			}
> 		}
> 		if opts.MemRequests != (resource.Quantity{}) {
> ```

Handle a request before a limit.
Move this up.

---

*[cmd/helm/installer/install.go, line 146 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KntJEa4t6xj9kOlpTqv:-KntJEa4t6xj9kOlpTqw:b-ia0163) ([raw file](https://github.com/kubernetes/helm/blob/10fd6f5c989e1284ef0e9276d989d5d6bc23828d/cmd/helm/installer/install.go#L146)):*
> ```Go
> 			requests[v1.ResourceMemory] = opts.MemRequests
> 		}
> 		resources = v1.ResourceRequirements{
> ```

Since you already have a `ResourceRequirements` object in hand, either assign to its fields here or eliminate the "resources" value declared above and return the struct literal directly here.
```go
resources.Requests = requests
resources.Limits = limits
```

---

*[cmd/helm/installer/install.go, line 211 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KntK2vqKYXtYk-ZYa_W:-KntK2vqKYXtYk-ZYa_X:bvlzawf) ([raw file](https://github.com/kubernetes/helm/blob/10fd6f5c989e1284ef0e9276d989d5d6bc23828d/cmd/helm/installer/install.go#L211)):*
> ```Go
> 		},
> 	}
> 	if resourcesEnabled {
> ```

Even if `generateLimits` returned a zero-valued `ResourceRequirements`, you could still assign it here. That would allow you to eliminate the second return value from `generateLimits`.

---

*[cmd/helm/installer/install_test.go, line 150 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KntKOmQtzXVNeSMTgJy:-KntKOmRF7ynaPaSG6Hm:bco1vef) ([raw file](https://github.com/kubernetes/helm/blob/10fd6f5c989e1284ef0e9276d989d5d6bc23828d/cmd/helm/installer/install_test.go#L150)):*
> ```Go
> 
> func TestDeploymentManifest_WithResourceLimits(t *testing.T) {
> ```

Remove blank line.

---

*[cmd/helm/installer/install_test.go, line 152 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KntKQkrJj4sSUBVBQWI:-KntKQksNQFVQA3xdoTQ:bi5ta62) ([raw file](https://github.com/kubernetes/helm/blob/10fd6f5c989e1284ef0e9276d989d5d6bc23828d/cmd/helm/installer/install_test.go#L152)):*
> ```Go
> 
> 	type test struct {
> 		cpuLimits   string
> ```

This and the three that follow should be singular.

---

*[cmd/helm/installer/install_test.go, line 175 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KntKUNU1zcD2hPFAAEl:-KntKUNU1zcD2hPFAAEm:bco1vef) ([raw file](https://github.com/kubernetes/helm/blob/10fd6f5c989e1284ef0e9276d989d5d6bc23828d/cmd/helm/installer/install_test.go#L175)):*
> ```Go
> 		{},
> 	}
> ```

Remove blank line.

---

*[cmd/helm/installer/install_test.go, line 230 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KntKYclF6EnZO9TuQIS:-KntKYclF6EnZO9TuQIT:bco1vef) ([raw file](https://github.com/kubernetes/helm/blob/10fd6f5c989e1284ef0e9276d989d5d6bc23828d/cmd/helm/installer/install_test.go#L230)):*
> ```Go
> 			t.Errorf("Expected memory requests %q, got %q", opts.MemRequests, got)
> 		}
> ```

Remove blank line.

---

*[cmd/helm/installer/options.go, line 76 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KntK_cxFEpBNbGemOZJ:-KntK_cy3OCN3oWgnJ0P:bexmoz0) ([raw file](https://github.com/kubernetes/helm/blob/10fd6f5c989e1284ef0e9276d989d5d6bc23828d/cmd/helm/installer/options.go#L76)):*
> ```Go
> 	EnableHostNetwork bool
> 
> 	//CPULimits is the cpu limits used to deploy tiller
> ```

Add a space between the comment delimiter and the first comment character.

---

*[cmd/helm/installer/options.go, line 76 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KntKiNt45hZzltxTQm1:-KntKiNt45hZzltxTQm2:bnc4vbe) ([raw file](https://github.com/kubernetes/helm/blob/10fd6f5c989e1284ef0e9276d989d5d6bc23828d/cmd/helm/installer/options.go#L76)):*
> cpu limits used to deploy tiller

That should read "CPU limit used when deploying Tiller." Punctuate the sentence.

---

*[cmd/helm/installer/options.go, line 76 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KntLDZm90g_KfdXktvm:-KntLDZncRrbtJA0pBxd:b-v8isoe) ([raw file](https://github.com/kubernetes/helm/blob/10fd6f5c989e1284ef0e9276d989d5d6bc23828d/cmd/helm/installer/options.go#L76)):*
> ```Go
> 	EnableHostNetwork bool
> 
> 	//CPULimits is the cpu limits used to deploy tiller
> ```

Address the request before the limit. Move this up.

---

*[cmd/helm/installer/options.go, line 77 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KntLGd7YzrFVA-8rWsR:-KntLGd7YzrFVA-8rWsS:b-ftfkz1) ([raw file](https://github.com/kubernetes/helm/blob/10fd6f5c989e1284ef0e9276d989d5d6bc23828d/cmd/helm/installer/options.go#L77)):*
> CPULimits

This should be singular: "CPULimit."

---

*[cmd/helm/installer/options.go, line 79 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KntKeKwUmvBSWrIpUhf:-KntKeKwUmvBSWrIpUhg:bexmoz0) ([raw file](https://github.com/kubernetes/helm/blob/10fd6f5c989e1284ef0e9276d989d5d6bc23828d/cmd/helm/installer/options.go#L79)):*
> ```Go
> 	CPULimits resource.Quantity
> 
> 	//CPURequests is the cpu request limits used to deploy tiller
> ```

Add a space between the comment delimiter and the first comment character.

---

*[cmd/helm/installer/options.go, line 79 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KntKrrA34edAy1_AGeL:-KntKrrByou2ZGVSzAdP:bbcn7kf) ([raw file](https://github.com/kubernetes/helm/blob/10fd6f5c989e1284ef0e9276d989d5d6bc23828d/cmd/helm/installer/options.go#L79)):*
> cpu request limits used to deploy tiller


That should read "CPU request used when deploying Tiller." Punctuate the sentence.

---

*[cmd/helm/installer/options.go, line 80 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KntLPGSN4JNugcOjTto:-KntLPGSN4JNugcOjTtp:b9k5g37) ([raw file](https://github.com/kubernetes/helm/blob/10fd6f5c989e1284ef0e9276d989d5d6bc23828d/cmd/helm/installer/options.go#L80)):*
> CPURequests


This should be singular: "CPURequest."

---

*[cmd/helm/installer/options.go, line 82 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KntKf8_zeic0sQkW_5L:-KntKf8_zeic0sQkW_5M:bexmoz0) ([raw file](https://github.com/kubernetes/helm/blob/10fd6f5c989e1284ef0e9276d989d5d6bc23828d/cmd/helm/installer/options.go#L82)):*
> ```Go
> 	CPURequests resource.Quantity
> 
> 	//MemLimits is the memory limits used to deploy tiller
> ```

Add a space between the comment delimiter and the first comment character.

---

*[cmd/helm/installer/options.go, line 82 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KntKyKOKGHQH-sKnu3J:-KntKyKOKGHQH-sKnu3K:b-h4ntvk) ([raw file](https://github.com/kubernetes/helm/blob/10fd6f5c989e1284ef0e9276d989d5d6bc23828d/cmd/helm/installer/options.go#L82)):*
> memory limits used to deploy tiller

That should read "memory limit used when deploying Tiller." Punctuate the sentence.

---

*[cmd/helm/installer/options.go, line 83 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KntLLvGPLo0TC2xPaKb:-KntLLvGPLo0TC2xPaKc:b2srsa9) ([raw file](https://github.com/kubernetes/helm/blob/10fd6f5c989e1284ef0e9276d989d5d6bc23828d/cmd/helm/installer/options.go#L83)):*
> MemLimits


This should be singular: "MemLimit."

---

*[cmd/helm/installer/options.go, line 85 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KntKfiqFuspKE3JEX95:-KntKfiqFuspKE3JEX96:bexmoz0) ([raw file](https://github.com/kubernetes/helm/blob/10fd6f5c989e1284ef0e9276d989d5d6bc23828d/cmd/helm/installer/options.go#L85)):*
> ```Go
> 	MemLimits resource.Quantity
> 
> 	//MemRequests is the memory request limits used to deploy tiller
> ```

Add a space between the comment delimiter and the first comment character.

---

*[cmd/helm/installer/options.go, line 85 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KntL2fK-eLuIzpRwOUr:-KntL2fLtlcq8gstsNas:b-554oo5) ([raw file](https://github.com/kubernetes/helm/blob/10fd6f5c989e1284ef0e9276d989d5d6bc23828d/cmd/helm/installer/options.go#L85)):*
> memory request limits used to deploy tiller

That should read "memory request used when deploying Tiller." Punctuate the sentence.

---

*[cmd/helm/installer/options.go, line 85 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KntL5x3ri0mqNP4Qv0K:-KntL5x3ri0mqNP4Qv0L:b-v8isoe) ([raw file](https://github.com/kubernetes/helm/blob/10fd6f5c989e1284ef0e9276d989d5d6bc23828d/cmd/helm/installer/options.go#L85)):*
> ```Go
> 	MemLimits resource.Quantity
> 
> 	//MemRequests is the memory request limits used to deploy tiller
> ```

Address the request before the limit. Move this up.

---

*[cmd/helm/installer/options.go, line 86 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KntLS_Q-POwaQUpfb45:-KntLS_Q-POwaQUpfb46:b-wsi44n) ([raw file](https://github.com/kubernetes/helm/blob/10fd6f5c989e1284ef0e9276d989d5d6bc23828d/cmd/helm/installer/options.go#L86)):*
> MemRequests


This should be singular: "MemRequest."

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632)*
<!-- Sent from Reviewable.io -->

seh commented:
Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 40 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632)*
<!-- Sent from Reviewable.io -->

ata18 committed:
Added resource limit/request flags to helm init

This PR adds the following 4 additional flags to be used with helm init
command so that appropriate CPU/Memory limits can be set for tiller
container.
--tiller-cpu-limit
--tiller-cpu-request
--tiller-memory-limit
--tiller-memory-request

To maintain backwards compatibility, when none of these flags are set,
no limitrange will be added. When tiller-cpu-limit is specified but
not the tiller-cpu-request, then tiller-cpu-request is set to same
value as the tiller-cpu-limit. But not vice-versa. Same applies to
tiller-memory-limits and tiller-memory-request.

Fixes: https://github.com/kubernetes/helm/issues/2135

at1012 commented:
@seh Thanks for the detailed review comments. I have addressed all of them. Could you please review this?

seh commented:
Reviewed 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 61 unresolved discussions.

---

*[cmd/helm/init.go, line 84 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-Ko90H5Jt5dhaKgPcpIt:-Ko90H5KtA4gxiSS9RWW:bw2vpla) ([raw file](https://github.com/kubernetes/helm/blob/b2ed3407432dc5478c6545c70473842f9e87f07d/cmd/helm/init.go#L84)):*
> ```Go
> 	serviceAccount string
> 	cpuLimit       string
> 	cpuRequest     string
> ```

Please move the limit-related variables up above the request-related variables, to better match the order in [`v1.ResourceRequirements`](https://github.com/kubernetes/kubernetes/blob/11b5956f7a37d61e5b264b461adf6b563a0c7b44/staging/src/k8s.io/api/core/v1/types.go#L1697-L1709).
That is, the order should be as follows:
- `cpuLimit`
- `memoryLimit`
- `cpuRequest
- `memoryRequest`

---

*[cmd/helm/init.go, line 86 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-Ko90_Q_4DcjN7JhxoQi:-Ko90_Q_4DcjN7JhxoQj:bgafpzi) ([raw file](https://github.com/kubernetes/helm/blob/b2ed3407432dc5478c6545c70473842f9e87f07d/cmd/helm/init.go#L86)):*
> ```Go
> 	cpuRequest     string
> 	memoryLimit    string
> 	memoryRequest  string
> ```

Likewise.

---

*[cmd/helm/init_test.go, line 323 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-Ko90eno6ac1NUwqGGBI:-Ko90eno6ac1NUwqGGBJ:bw2vpla) ([raw file](https://github.com/kubernetes/helm/blob/b2ed3407432dc5478c6545c70473842f9e87f07d/cmd/helm/init_test.go#L323)):*
> ```Go
> 		namespace:     v1.NamespaceDefault,
> 		cpuLimit:      "2",
> 		cpuRequest:    "100m",
> ```

Please move the limit-related variables up above the request-related variables, to better match the order in [`v1.ResourceRequirements`](https://github.com/kubernetes/kubernetes/blob/11b5956f7a37d61e5b264b461adf6b563a0c7b44/staging/src/k8s.io/api/core/v1/types.go#L1697-L1709).
That is, the order should be as follows:
- `cpuLimit`
- `memoryLimit`
- `cpuRequest
- `memoryRequest`

---

*[cmd/helm/installer/install.go, line 127 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-Ko92MpD3R-AtNNs8rZa:-Ko92MpEmfk6tce4syf7:b-44xsu8) ([raw file](https://github.com/kubernetes/helm/blob/b2ed3407432dc5478c6545c70473842f9e87f07d/cmd/helm/installer/install.go#L127)):*
> ```Go
> 		requests[v1.ResourceCPU] = opts.CPURequest
> 	}
> 	if opts.CPULimit != (resource.Quantity{}) {
> ```

I apologize; in my last review pass, I asked you to handle the requests before the limits, but having reviewed more of the Kubernetes type definitions and the documentation, I see that my preference was misaligned; instead, we should handle limits first, then requests, and for consistency within those groups, handle CPU before memory. Hence, the order I'd prefer is as follows:
- `cpuLimit`
- `memoryLimit`
- `cpuRequest`
- `memoryRequest`

---

*[cmd/helm/installer/install.go, line 129 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-Ko934zdGvQ3Fq3S9mzl:-Ko934zdGvQ3Fq3S9mzm:b-vqy1q6) ([raw file](https://github.com/kubernetes/helm/blob/b2ed3407432dc5478c6545c70473842f9e87f07d/cmd/helm/installer/install.go#L129)):*
> ```Go
> 	if opts.CPULimit != (resource.Quantity{}) {
> 		limits[v1.ResourceCPU] = opts.CPULimit
> 		if opts.CPURequest == (resource.Quantity{}) {
> ```

Why do force the request to equal the limit? That gives us the pod the "Guaranteed" QoS class, but we're cutting the "LimitRanger" admission control plugin out of the picture here. I'd prefer to honor the user's specified options: If he expresses no request, we pass the pod spec through with no expressed request.

---

*[cmd/helm/installer/install.go, line 129 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-Ko94oX7L7AI5PoH1iY3:-Ko94oX8vSlW4rVeH-vq:b-g6whsn) ([raw file](https://github.com/kubernetes/helm/blob/b2ed3407432dc5478c6545c70473842f9e87f07d/cmd/helm/installer/install.go#L129)):*
> ```Go
> 	if opts.CPULimit != (resource.Quantity{}) {
> 		limits[v1.ResourceCPU] = opts.CPULimit
> 		if opts.CPURequest == (resource.Quantity{}) {
> ```

Did you mean to check whether `opts.CPURequest` is zero? If so, use [`(*Quantity).IsZero`](https://github.com/kubernetes/apimachinery/blob/e42df1bc0a097d6d62d230b398b84096238ea43c/pkg/api/resource/quantity.go#L497-L503) instead.

---

*[cmd/helm/installer/install.go, line 133 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-Ko95AK_vlY9nTzmtTMy:-Ko95AK_vlY9nTzmtTMz:byrvjho) ([raw file](https://github.com/kubernetes/helm/blob/b2ed3407432dc5478c6545c70473842f9e87f07d/cmd/helm/installer/install.go#L133)):*
> ```Go
> 		}
> 	}
> 	if opts.MemoryRequest != (resource.Quantity{}) {
> ```

Did you mean to check whether `opts.MemoryRequest` is zero? If so, use [`(*Quantity).IsZero`](https://github.com/kubernetes/apimachinery/blob/e42df1bc0a097d6d62d230b398b84096238ea43c/pkg/api/resource/quantity.go#L497-L503) instead.

---

*[cmd/helm/installer/install.go, line 136 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-Ko95D2SCWpCOKoBSrXq:-Ko95D2SCWpCOKoBSrXr:b-hg9fzc) ([raw file](https://github.com/kubernetes/helm/blob/b2ed3407432dc5478c6545c70473842f9e87f07d/cmd/helm/installer/install.go#L136)):*
> ```Go
> 		requests[v1.ResourceMemory] = opts.MemoryRequest
> 	}
> 	if opts.MemoryLimit != (resource.Quantity{}) {
> ```

Did you mean to check whether `opts.MemoryLimit` is zero? If so, use [`(*Quantity).IsZero`](https://github.com/kubernetes/apimachinery/blob/e42df1bc0a097d6d62d230b398b84096238ea43c/pkg/api/resource/quantity.go#L497-L503) instead.

---

*[cmd/helm/installer/install.go, line 138 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-Ko93pjxDL8WBTCkqOEy:-Ko93AmgZcwYPE9v_Z8U:b-3huro4) ([raw file](https://github.com/kubernetes/helm/blob/b2ed3407432dc5478c6545c70473842f9e87f07d/cmd/helm/installer/install.go#L138)):*
> ```Go
> 	if opts.MemoryLimit != (resource.Quantity{}) {
> 		limits[v1.ResourceMemory] = opts.MemoryLimit
> 		if opts.MemoryRequest == (resource.Quantity{}) {
> ```

Likewise, let's not cut the "LimitRanger" admission control plugin out here.

---

*[cmd/helm/installer/install.go, line 138 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-Ko95FsdIBw3-XtbVzEz:-Ko95FsdIBw3-XtbVzF-:byrvjho) ([raw file](https://github.com/kubernetes/helm/blob/b2ed3407432dc5478c6545c70473842f9e87f07d/cmd/helm/installer/install.go#L138)):*
> ```Go
> 	if opts.MemoryLimit != (resource.Quantity{}) {
> 		limits[v1.ResourceMemory] = opts.MemoryLimit
> 		if opts.MemoryRequest == (resource.Quantity{}) {
> ```

Did you mean to check whether `opts.MemoryRequest` is zero? If so, use [`(*Quantity).IsZero`](https://github.com/kubernetes/apimachinery/blob/e42df1bc0a097d6d62d230b398b84096238ea43c/pkg/api/resource/quantity.go#L497-L503) instead.

---

*[cmd/helm/installer/install_test.go, line 150 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-Ko95ZlMHlhQqzXfTB77:-Ko95ZlMHlhQqzXfTB78:b4bngtm) ([raw file](https://github.com/kubernetes/helm/blob/b2ed3407432dc5478c6545c70473842f9e87f07d/cmd/helm/installer/install_test.go#L150)):*
> ```Go
> 
> func TestDeploymentManifest_WithResourceLimits(t *testing.T) {
> 	type test struct {
> ```

You could eliminate this named type:
```go
tests := []struct{
	cpuLimit   string
	memoryLimit   string
	cpuRequest string
	memoryRequest string
}{
	{cpuLimit: "2", memoryLimit: "1Gi", cpuRequest: "100m", memoryRequest: "500Mi"},
}
```

---

*[cmd/helm/installer/install_test.go, line 202 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-Ko96JBfPMiM0OZBj86k:-Ko96JBgt3hdETp1Lxbz:b-wb3t09) ([raw file](https://github.com/kubernetes/helm/blob/b2ed3407432dc5478c6545c70473842f9e87f07d/cmd/helm/installer/install_test.go#L202)):*
> cpu limits

That should be "CPU limit."

---

*[cmd/helm/installer/install_test.go, line 205 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-Ko96N98uqkmIQ6M1TPm:-Ko96N98uqkmIQ6M1TPn:b-kiype4) ([raw file](https://github.com/kubernetes/helm/blob/b2ed3407432dc5478c6545c70473842f9e87f07d/cmd/helm/installer/install_test.go#L205)):*
> limits

That should be singular: "limit."

---

*[cmd/helm/installer/install_test.go, line 207 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-Ko96oKTpoJ3UlfJB-h0:-Ko96oKUNqCDm8-_58Md:biek1vh) ([raw file](https://github.com/kubernetes/helm/blob/b2ed3407432dc5478c6545c70473842f9e87f07d/cmd/helm/installer/install_test.go#L207)):*
> && got != opts.CPULimit

As suggested above, we should not set the request to equal the limit.

---

*[cmd/helm/installer/install_test.go, line 208 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-Ko96Vl_DoMsDEIp1Kbe:-Ko96Vl_DoMsDEIp1Kbf:b-sh8m44) ([raw file](https://github.com/kubernetes/helm/blob/b2ed3407432dc5478c6545c70473842f9e87f07d/cmd/helm/installer/install_test.go#L208)):*
> cpu requests

That should be singular: "CPU request."

---

*[cmd/helm/installer/install_test.go, line 210 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-Ko96jChFkIk0uMCsp0J:-Ko96jChFkIk0uMCsp0K:bwftppg) ([raw file](https://github.com/kubernetes/helm/blob/b2ed3407432dc5478c6545c70473842f9e87f07d/cmd/helm/installer/install_test.go#L210)):*
> && got != opts.MemoryLimit

As suggested above, we should not set the request to equal the limit.

---

*[cmd/helm/installer/install_test.go, line 211 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-Ko96_Dee4fvEEUw83S4:-Ko96_Dee4fvEEUw83S5:b-frbkro) ([raw file](https://github.com/kubernetes/helm/blob/b2ed3407432dc5478c6545c70473842f9e87f07d/cmd/helm/installer/install_test.go#L211)):*
> requests

That should be singular: "request."

---

*[cmd/helm/installer/options.go, line 76 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-Ko96uxe89Demcw_WI70:-Ko96uxe89Demcw_WI71:b-xurdvt) ([raw file](https://github.com/kubernetes/helm/blob/b2ed3407432dc5478c6545c70473842f9e87f07d/cmd/helm/installer/options.go#L76)):*
> tiller

Per #2601, please write that as the proper noun "Tiller."

---

*[cmd/helm/installer/options.go, line 79 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-Ko979Qoe63LVBVqj0jI:-Ko979Qoe63LVBVqj0jJ:b-xurdvt) ([raw file](https://github.com/kubernetes/helm/blob/b2ed3407432dc5478c6545c70473842f9e87f07d/cmd/helm/installer/options.go#L79)):*
> tiller

Per #2601, please write that as the proper noun "Tiller."

---

*[cmd/helm/installer/options.go, line 82 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-Ko97BWFbZFdKZ_L7uJJ:-Ko97BWFbZFdKZ_L7uJK:b-xurdvt) ([raw file](https://github.com/kubernetes/helm/blob/b2ed3407432dc5478c6545c70473842f9e87f07d/cmd/helm/installer/options.go#L82)):*
> tiller

Per #2601, please write that as the proper noun "Tiller."

---

*[cmd/helm/installer/options.go, line 85 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-Ko97DMONzW0dGRgOeoG:-Ko97DMONzW0dGRgOeoH:b-xurdvt) ([raw file](https://github.com/kubernetes/helm/blob/b2ed3407432dc5478c6545c70473842f9e87f07d/cmd/helm/installer/options.go#L85)):*
> tiller

Per #2601, please write that as the proper noun "Tiller."

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632)*
<!-- Sent from Reviewable.io -->

at1012 commented:
@seh Addressed review comments.

seh commented:
Just a few nits remain (capitalization and parentheses). I expect this will be the last round.

---

Reviewed 5 of 5 files at r3.
Review status: all files reviewed at latest revision, 68 unresolved discussions.

---

*[cmd/helm/installer/install.go, line 124 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KoEOsuvh0_ZCH9JZfFL:-KoEOsuvh0_ZCH9JZfFM:b5e8ifr) ([raw file](https://github.com/kubernetes/helm/blob/ffa06a8d8caaf75b9910b569fcdc9b4aca2be76b/cmd/helm/installer/install.go#L124)):*
> (opts.CPULimit)

Those parentheses appear to be spurious. Does the following compile as well?
```go
if !opts.CPULimit.IsZero() {
```

---

*[cmd/helm/installer/install.go, line 127 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KoEPAGQRl1t_isANJEo:-KoEPAGQRl1t_isANJEp:b-dqywdp) ([raw file](https://github.com/kubernetes/helm/blob/ffa06a8d8caaf75b9910b569fcdc9b4aca2be76b/cmd/helm/installer/install.go#L127)):*
> (opts.MemoryLimit)

Those parentheses appear to be spurious. Does the following compile as well?
```go
if !opts.MemoryLimit.IsZero() {
```

---

*[cmd/helm/installer/install.go, line 130 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KoEPEvLN4p3LoFLa5zg:-KoEPEvLN4p3LoFLa5zh:b-7xq0kp) ([raw file](https://github.com/kubernetes/helm/blob/ffa06a8d8caaf75b9910b569fcdc9b4aca2be76b/cmd/helm/installer/install.go#L130)):*
> (opts.CPURequest)

Those parentheses appear to be spurious. Does the following compile as well?
```go
if !opts.CPURequest.IsZero() {
```

---

*[cmd/helm/installer/install.go, line 133 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KoEPIldDOxeLpS2gtRu:-KoEPIldDOxeLpS2gtRv:b5q117v) ([raw file](https://github.com/kubernetes/helm/blob/ffa06a8d8caaf75b9910b569fcdc9b4aca2be76b/cmd/helm/installer/install.go#L133)):*
> (opts.MemoryRequest)

Those parentheses appear to be spurious. Does the following compile as well?
```go
if !opts.MemoryRequest.IsZero() {
```

---

*[cmd/helm/installer/install_test.go, line 197 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KoEP_-ax_PXo2ueGLE9:-KoEP_-ax_PXo2ueGLEA:b75sow4) ([raw file](https://github.com/kubernetes/helm/blob/ffa06a8d8caaf75b9910b569fcdc9b4aca2be76b/cmd/helm/installer/install_test.go#L197)):*
> verify

Capitalize the first word in the sentence.

---

*[cmd/helm/installer/install_test.go, line 197 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KoEPedLVgDr_KQBbNXZ:-KoEPedLVgDr_KQBbNX_:b-mi928q) ([raw file](https://github.com/kubernetes/helm/blob/ffa06a8d8caaf75b9910b569fcdc9b4aca2be76b/cmd/helm/installer/install_test.go#L197)):*
> deployment

Capitalize the kind "Deployment."

---

*[cmd/helm/installer/install_test.go, line 197 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-KoEPicvj8EWGJ__pXSG:-KoEPicvj8EWGJ__pXSH:b-lgq8j) ([raw file](https://github.com/kubernetes/helm/blob/ffa06a8d8caaf75b9910b569fcdc9b4aca2be76b/cmd/helm/installer/install_test.go#L197)):*
> cpu

Capitalize "CPU."

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-:-KoEQ9V9wnKV4aGe7GVE:b-a6j6mo)*
<!-- Sent from Reviewable.io -->

at1012 commented:
@seh, I have addressed them as well. Thanks!

seh commented:
Thank you, @at1012. This LGTM.
I'd like final confirmation from @technosophos that this new capability doesn't conflict with any intentions he may have.

---

Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2632#-:-KoHf3tyG1X2VXSdkyT4:b-7rgafg)*
<!-- Sent from Reviewable.io -->

technosophos commented:
This one might conflict with one of the outstanding PRs @jascott1 is working on (which provided a general-purpose flag for setting params like this). We need to decide on whether we want to add these as a convenience.

jascott1 commented:
There were some strong opinions not to add a flag for every property of init manifest so we created `--set` and `--output` to allow setting individual values and dumping the whole manifest to a tool like `jq` respectively.

at1012 commented:
That sounds reasonable.

seh commented:
@at1012, I apologize for dragging you through this review, not knowing that the project intends to address this need in a different way.

I wonder, though, whether a generic value setting approach will be able to parse these resource values correctly, and detect when they're equal to zero. I don't think that means that we should an ever-expanding set of command-line flags to cover every possible facet of Tiller's Kubernetes objects (that's a losing proposition, as more and more fields arrive and become relevant to Tiller in each Kubernetes release), but I am also uncertain whether we'll be able to handle all the different kinds of values generically.

ata18 committed:
Updated docs

at1012 commented:
@seh I learned a few things, so not a problem. @technosophos @jascott1 I agree to Steven's point. I am not sure if the generic approach will work in this particular case. So I'm keeping this PR around. If determined, that it is a repeated work, feel free to close this.

thomastaylor312 commented:
@technosophos, @seh, @jascott1 So did we want to keep this open for reference or close it for now?

seh commented:
I don't recall whether we settled on favoring a more generic means of setting manifest fields like these. What I haven't tested yet is to see whether it is indeed feasible to set these four values (using strings that will be converted to `resource.Quantity` values).

Related: issue #2299 and PR #2557.

thomastaylor312 commented:
I believe this has been addressed with the `--override` flag. Closing
